### PR TITLE
[Merged by Bors] - feat (RingTheory/HahnSeries/Summable) add pointwise product operation

### PR DESCRIPTION
**Pull request by ScottCarnahan**
This PR defines a pointwise product operation on summable families of Hahn series, and shows that the formal sum of the pointwise product is equal to the product of formal sums.  This is preparation for defining algebra homomorphisms from power series rings.

---

[Open in Gitpod](https://gitpod.io/from-referrer/)

### CODE REVIEW

File: Mathlib/RingTheory/HahnSeries/Summable.lean [hunk 1]
```diff
@@ -267,7 +267,24 @@ end AddCommGroup
...
+        Pi.smul_apply r t i ▸ Function.support_const_smul_subset r _)
+      finite_co_support' := by
+        intro g
+        refine (t.finite_co_support g).subset ?_
+        intro i hi
+        simp only [Pi.smul_apply, smul_coeff, ne_eq, Set.mem_setOf_eq] at hi
+        simp only [Function.mem_support, ne_eq]
+        exact right_ne_zero_of_smul hi
+    }
+  ⟩
```
Comment (kim-em) [inline Mathlib/RingTheory/HahnSeries/Summable.lean:285]: Very minor, but I think the general consensus in Mathlib is that closing parens don't deserve their own line.

File: Mathlib/RingTheory/HahnSeries/Summable.lean [hunk 2]
```diff
@@ -427,6 +444,40 @@ theorem hsum_sub {R : Type*} [Ring R] {s t : SummableFamily Γ R α} :
...
+  isPWO_iUnion_support_prod_smul hs ht
+
+theorem finite_co_support_prod_mul (s : SummableFamily Γ R α)
+    (t : SummableFamily Γ R β) (g : Γ) :
+    Finite {(a : α × β) | ((fun (a : α × β) ↦ (s a.1 * t a.2)) a).coeff g ≠ 0} :=
+  finite_co_support_prod_smul s t g
+
+/-- A summable family given by pointwise multiplication of a pair of summable families. -/
+@[simps]
+def FamilyMul (s : SummableFamily Γ R α) (t : SummableFamily Γ R β) :
```
Comment (kim-em) [inline Mathlib/RingTheory/HahnSeries/Summable.lean:459]: I don't think this is the right name. Perhaps `SummableFamily.mul`?

File: Mathlib/RingTheory/HahnSeries/Summable.lean [hunk 3]
```diff
@@ -427,6 +444,40 @@ theorem hsum_sub {R : Type*} [Ring R] {s t : SummableFamily Γ R α} :
...
+    (SummableFamily Γ R (α × β)) where
+  toFun a := s (a.1) * t (a.2)
+  isPWO_iUnion_support' :=
+    isPWO_iUnion_support_prod_mul s.isPWO_iUnion_support t.isPWO_iUnion_support
+  finite_co_support' g := finite_co_support_prod_mul s t g
+
+theorem familymul_eq_familysmul {β : Type*} (s : SummableFamily Γ R α) (t : SummableFamily Γ R β) :
+    FamilyMul s t = FamilySMul s t :=
```
Comment (kim-em) [inline Mathlib/RingTheory/HahnSeries/Summable.lean:467]: Hmm, I see we've maybe already gone wrong with names... Not too late to fix! :-)

**Review by jcommelin — COMMENTED** [review]
Thanks :tada:

bors merge

**Pull request closed without merging**